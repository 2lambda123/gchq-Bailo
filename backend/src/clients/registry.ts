import fetch from 'node-fetch'

import { getHttpsAgent } from '../services/v2/http.js'
import config from '../utils/v2/config.js'
import { InternalError } from '../utils/v2/error.js'

interface RepoRef {
  namespace: string
  image: string
}

const registry = config.registry.connection.internal

const agent = getHttpsAgent({
  rejectUnauthorized: !config.registry.connection.insecure,
})

async function registryRequest(token: string, endpoint: string) {
  let res
  try {
    res = await fetch(`${registry}/v2/${endpoint}`, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
      agent,
    })
  } catch (err) {
    throw InternalError('Unable to communicate with the registry.', { err })
  }

  if (!res.ok) {
    throw InternalError('Non-200 response returned by the registry.', {
      status: res.status,
      statusText: res.statusText,
      body: await res.text(),
    })
  }

  return await res.json()
}

// Currently limited to a maximum 100 image names
type ListModelReposResponse = { repositories: Array<string> }
export async function listModelRepos(token: string, modelId: string) {
  const responseBody = await registryRequest(token, `_catalog?n=100&last=${modelId}`)
  if (!isListModelReposResponse(responseBody)) {
    throw InternalError('Unrecognised response body when listing model repositories.', { responseBody: responseBody })
  }

  const filteredRepos = responseBody.repositories.filter((repo) => repo.startsWith(`${modelId}/`))
  return filteredRepos
}

function isListModelReposResponse(resp: unknown): resp is ListModelReposResponse {
  if (typeof resp !== 'object' || resp === null) {
    return false
  }
  if (!('repositories' in resp) || !Array.isArray(resp.repositories)) {
    return false
  }
  return true
}

type ListImageTagResponse = { tags: Array<string> }
export async function listImageTags(token: string, imageRef: RepoRef) {
  const repo = `${imageRef.namespace}/${imageRef.image}`

  const responseBody = await registryRequest(token, `${repo}/tags/list`)
  if (!isListImageTagResponse(responseBody)) {
    throw InternalError('Unrecognised response body when listing image tags.', { responseBody: responseBody })
  }
  return responseBody.tags
}

function isListImageTagResponse(resp: unknown): resp is ListImageTagResponse {
  if (typeof resp !== 'object' || resp === null) {
    return false
  }
  if (!('tags' in resp) || !Array.isArray(resp.tags)) {
    return false
  }
  return true
}
