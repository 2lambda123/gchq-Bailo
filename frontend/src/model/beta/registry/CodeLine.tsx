import { ContentCopy } from '@mui/icons-material'
import CodeIcon from '@mui/icons-material/Code'
import { Box, IconButton, Stack, Tooltip, Typography } from '@mui/material'
import { useTheme } from '@mui/material/styles'
import { ReactElement } from 'react'
import useNotification from 'src/common/Snackbar'

interface CodeLineProps {
  line: string
  icon?: ReactElement
}

export default function CodeLine({ line, icon = <CodeIcon /> }: CodeLineProps) {
  const theme = useTheme()
  const sendNotification = useNotification()

  function handleButtonClick() {
    navigator.clipboard.writeText(line)
    sendNotification({ variant: 'success', msg: 'Copied to clipboard' })
  }

  return (
    <Stack direction='row' spacing={1}>
      <Box
        sx={{ backgroundColor: theme.palette.container.main, p: 1, borderRadius: 2, width: '100%' }}
        component='div'
        role='button'
        tabIndex={0}
      >
        <Stack direction='row' spacing={2} alignItems='center'>
          {icon}
          <Typography>{line}</Typography>
        </Stack>
      </Box>
      <Tooltip title='Copy to clipboard' arrow>
        <IconButton onClick={() => handleButtonClick()} aria-label='Copy text to clipboard'>
          <ContentCopy />
        </IconButton>
      </Tooltip>
    </Stack>
  )
}
