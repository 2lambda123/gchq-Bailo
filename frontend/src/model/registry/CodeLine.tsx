import CodeIcon from '@mui/icons-material/Code'
import { Box, Stack, Typography } from '@mui/material'
import { useTheme } from '@mui/material/styles'
import { ReactElement } from 'react'
import CopyToClipboardButton from 'src/common/CopyToClipboardButton'

interface CodeLineProps {
  line: string
  icon?: ReactElement
}

export default function CodeLine({ line, icon = <CodeIcon /> }: CodeLineProps) {
  const theme = useTheme()

  return (
    <Stack direction='row' spacing={1}>
      <Box
        sx={{ backgroundColor: theme.palette.container.main, p: 1, borderRadius: 2, width: '100%' }}
        component='div'
        role='button'
        tabIndex={0}
      >
        <Stack direction='row' spacing={2} alignItems='center'>
          {icon}
          <Typography>{line}</Typography>
        </Stack>
      </Box>
      <CopyToClipboardButton
        textToCopy={line}
        notificationText='Copied to clipboard'
        ariaLabel='Copy text to clipboard'
      />
    </Stack>
  )
}
