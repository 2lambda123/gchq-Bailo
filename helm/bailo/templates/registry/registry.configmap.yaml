---
apiVersion: v1
kind: ConfigMap
metadata:
  name: {{ include "bailo.fullname" . }}-registry
  labels:
    {{- include "bailo.labels" . | nindent 4 }}
data:
  registry.conf: |
    version: 0.1
    log:
      level: debug
      fields:
        service: registry
    storage:
      cache:
        blobdescriptor: inmemory
      s3:
        regionendpoint: http://{{ include "bailo.minio.host" . }}:{{ .Values.minio.service.ports.api }}
        bucket: {{ .Values.minio.registryBucket }}
        secure: false
        region: ignored-for-minio
      redirect:
        disable: true
      delete:
        enabled: true
      maintenance:
        uploadpurging:
          enabled: true
          age: 168h
          interval: 24h
          dryrun: false
        readonly:
          enabled: false
    http:
      addr: :{{ .Values.registry.port }}
      relativeurls: true
      headers:
        X-Content-Type-Options: [nosniff]
    health:
      storagedriver:
        enabled: true
        interval: 10s
        threshold: 3
