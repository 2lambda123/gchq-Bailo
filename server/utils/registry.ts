import https from 'https'
import config from 'config'
import { getAccessToken } from '../routes/v1/registryAuth'
import logger from './logger'

interface ImageRef {
  namespace: string
  image: string
  version: string
}

const httpsAgent = new https.Agent({
  rejectUnauthorized: !config.get('registry.insecure'),
})

export async function makeRegistryRequest({
  endpoint,
  authorisation,
  headers,
  metadata,
  method,
  json = true,
}: {
  endpoint: string
  authorisation: string
  method?: string
  headers?: any
  metadata?: any
  json?: boolean
}) {
  const registry = `${config.get('registry.protocol')}://${config.get('registry.host')}/v2`

  console.log(`${registry}${endpoint}`, {
    ...metadata,
    ...(method ? { method } : {}),
    headers: {
      ...headers,
      Authorization: 'authorisation bearer',
    },
  })

  return fetch(`${registry}${endpoint}`, {
    ...metadata,
    ...(method ? {} : { method }),
    headers: {
      ...headers,
      Authorization: authorisation,
    },
    agent: httpsAgent,
  } as RequestInit).then((res: any) => {
    logger.info(
      {
        status: res.status,
        endpoint,
      },
      'Made registry request'
    )

    if (res.status >= 400) {
      throw new Error(`Invalid status response: ${res.status}`)
    }

    if (json) {
      return res.json()
    }

    return res
  })
}

export async function copyDockerImage(from: ImageRef, to: ImageRef, log: (level: string, message: string) => void) {
  const token = await getAccessToken({ id: 'admin', _id: 'admin' }, [
    { type: 'repository', name: `${from.namespace}/${from.image}`, actions: ['pull'] },
    { type: 'repository', name: `${to.namespace}/${to.image}`, actions: ['push', 'pull'] },
  ])
  const authorisation = `Bearer ${token}`

  const manifest = await makeRegistryRequest({
    endpoint: `/${from.namespace}/${from.image}/manifests/${from.version}`,
    authorisation,
    headers: {
      Accept: 'application/vnd.docker.distribution.manifest.v2+json',
    },
  })

  console.log(manifest)

  await Promise.all(
    manifest.layers.map(async (layer: any) => {
      await makeRegistryRequest({
        endpoint: `/${to.namespace}/${to.image}/blobs/uploads/?mount=${layer.digest}&from=${from.namespace}/${from.image}`,
        authorisation,
        json: false,
        method: 'POST',
      })

      log('info', `Copied layer ${layer.digest}`)
    })
  )

  await makeRegistryRequest({
    endpoint: `/${to.namespace}/${to.image}/blobs/uploads/?mount=${manifest.config.digest}&from=${from.namespace}/${from.image}`,
    method: 'POST',
    authorisation,
  })

  log('info', 'Copied manifest to new repository')

  await makeRegistryRequest({
    endpoint: `/${to.namespace}/${to.image}/manifests/${to.version}`,
    method: 'PUT',
    authorisation,
    headers: {
      'Content-Type': 'application/vnd.docker.distribution.manifest.v2+json',
    },
  })

  log('info', 'Finalised new manifest')
}
