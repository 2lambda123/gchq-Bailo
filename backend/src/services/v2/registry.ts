import { listImageTags, listModelRepos } from '../../clients/registry.js'
import { ImageAction } from '../../connectors/v2/authorisation/base.js'
import authorisation from '../../connectors/v2/authorisation/index.js'
import { UserDoc } from '../../models/v2/User.js'
import { Action, getAccessToken } from '../../routes/v1/registryAuth.js'
import { Forbidden } from '../../utils/v2/error.js'
import { getModelById } from './model.js'

export async function listModelImages(user: UserDoc, modelId: string) {
  const model = await getModelById(user, modelId)

  const auth = await authorisation.image(user, model, {
    type: 'repository',
    name: modelId,
    actions: [ImageAction.List as Action],
  })
  if (!auth.success) {
    throw Forbidden(auth.info, { userDn: user.dn, modelId })
  }

  const registryToken = await getAccessToken({ id: user.dn, _id: user.dn }, [
    { type: 'registry', class: '', name: 'catalog', actions: ['*'] },
  ])
  const repos = await listModelRepos(registryToken, modelId)
  const versions = await Promise.all(
    repos.map(async (repo) => {
      const [namespace, image] = repo.split('/')
      const repositoryToken = await getAccessToken({ id: user.dn, _id: user.dn }, [
        { type: 'repository', class: '', name: repo, actions: ['pull'] },
      ])
      return { repository: namespace, name: image, tags: await listImageTags(repositoryToken, { namespace, image }) }
    }),
  )

  return versions
}
