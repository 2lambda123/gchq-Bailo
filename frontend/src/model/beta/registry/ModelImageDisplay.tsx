import { Box, Stack, Typography } from '@mui/material'
import { useTheme } from '@mui/material/styles'
import { useGetUiConfig } from 'actions/uiConfig'
import { useMemo } from 'react'
import Loading from 'src/common/Loading'
import MessageAlert from 'src/MessageAlert'
import CodeLine from 'src/model/beta/registry/CodeLine'

import { ModelImage } from '../../../../types/interfaces'

type ModelImageDisplayProps = {
  modelImage: ModelImage
}

export default function ModelImageDisplay({ modelImage }: ModelImageDisplayProps) {
  const theme = useTheme()

  const { uiConfig, isUiConfigLoading, isUiConfigError } = useGetUiConfig()

  const modelImageTags = useMemo(
    () =>
      uiConfig &&
      modelImage.tags.map((imageTag) => (
        <CodeLine
          key={`${modelImage.name}-${imageTag}`}
          line={`${uiConfig.registry.host}/${modelImage.repository}/${modelImage.name}:${imageTag}`}
        />
      )),
    [modelImage, uiConfig],
  )

  if (isUiConfigError) {
    return <MessageAlert message={isUiConfigError.info.message} severity='error' />
  }

  return (
    <>
      {isUiConfigLoading && <Loading />}
      <Box
        sx={{
          borderWidth: '1px',
          borderStyle: 'solid',
          borderColor: theme.palette.primary.main,
          width: '100%',
          borderRadius: 4,
          p: 2,
        }}
      >
        <Stack direction='column' spacing={1}>
          <Typography component='h2' variant='h6' color='primary'>
            {modelImage.name}
          </Typography>
          {modelImageTags}
        </Stack>
      </Box>
    </>
  )
}
