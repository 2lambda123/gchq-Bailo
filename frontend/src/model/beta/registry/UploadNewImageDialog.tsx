import ContentCopy from '@mui/icons-material/ContentCopy'
import {
  Box,
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  Divider,
  IconButton,
  Stack,
  Tooltip,
  Typography,
} from '@mui/material'
import { useTheme } from '@mui/material/styles'
import { useGetUiConfig } from 'actions/uiConfig'
import { useGetCurrentUser } from 'actions/user'
import { useState } from 'react'
import shellEscape from 'shell-escape'
import Loading from 'src/common/Loading'
import MessageAlert from 'src/MessageAlert'
import CodeLine from 'src/model/beta/registry/CodeLine'
import { ModelInterface } from 'types/v2/types'
import { getErrorMessage } from 'utils/fetcher'

interface UploadModelImageDialogProps {
  open: boolean
  handleClose: () => void
  model: ModelInterface
}

export default function UploadModelImageDialog({ open, handleClose, model }: UploadModelImageDialogProps) {
  const { uiConfig, isUiConfigLoading, isUiConfigError } = useGetUiConfig()
  const { currentUser, isCurrentUserLoading, isCurrentUserError } = useGetCurrentUser()

  const theme = useTheme()

  const [displayToken, setDisplayToken] = useState(false)
  const [displayedToken, setDisplayedToken] = useState('')
  const [tokenErrorText, setTokenErrorText] = useState('')

  const regenerateToken = async () => {
    const res = await fetch('/api/v1/user/token', {
      method: 'POST',
    })

    if (!res.ok) {
      setTokenErrorText(await getErrorMessage(res))
      return
    }

    let token

    try {
      const body = await res.json()
      token = body.token
    } catch (error) {
      setTokenErrorText('Recieved invalid response from server.')
    }

    return token
  }

  const showToken = async () => {
    const token = await regenerateToken()
    setDisplayedToken(token)
    setDisplayToken(true)
  }

  const copyTokenOnClick = async () => {
    const token = await regenerateToken()
    if (token) {
      setDisplayedToken(token)
      navigator.clipboard.writeText(token)
    }
  }

  if (isUiConfigError) {
    return <MessageAlert message={isUiConfigError.info.message} severity='error' />
  }

  if (isCurrentUserError) {
    return <MessageAlert message={isCurrentUserError.info.message} severity='error' />
  }

  return (
    <>
      {(isUiConfigLoading || isCurrentUserLoading) && <Loading />}
      {uiConfig && currentUser && (
        <Dialog open={open} onClose={handleClose}>
          <DialogTitle color='primary'>Pushing an Image for this Model</DialogTitle>
          <DialogContent>
            <Stack spacing={2}>
              <Typography fontWeight='bold'>User authentication token</Typography>
              <Typography>Use the token below to authenticate when you try and run the docker login command</Typography>
              <Stack direction='row'>
                <Button sx={{ mr: 2 }} variant='outlined' onClick={showToken} data-test='showTokenButton'>
                  Regenerate Token
                </Button>
                <Box
                  sx={{
                    backgroundColor: theme.palette.container.main,
                    px: 2,
                    display: 'flex',
                    mr: 1,
                  }}
                >
                  <Box component={Stack} direction='column' justifyContent='center'>
                    <Typography variant='body1' data-test='dockerPassword'>
                      {displayToken ? displayedToken : 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxx'}
                    </Typography>
                  </Box>
                </Box>
                <Tooltip title='Regenerate & Copy to clipboard'>
                  <IconButton onClick={copyTokenOnClick} aria-label='regenerate token and copy to clipboard'>
                    <ContentCopy />
                  </IconButton>
                </Tooltip>
              </Stack>
              <Typography variant='caption' color={theme.palette.error.main}>
                {tokenErrorText}
              </Typography>
              <Stack spacing={1}>
                <Typography fontWeight='bold'>Logging in</Typography>
                <Stack spacing={2}>
                  <CodeLine line={`docker login ${uiConfig.registry.host} -u ${shellEscape([currentUser.dn])}`} />
                </Stack>
              </Stack>
              <Stack spacing={1}>
                <Typography fontWeight='bold'>Pushing an image to the registry</Typography>
                <Stack spacing={2}>
                  <CodeLine line={`docker tag <image> ${window.location.host}/${model.id}/<name>:<tag>`} />
                  <CodeLine line={`docker push ${window.location.host}/${model.id}/<name>:<tag>`} />
                </Stack>
              </Stack>
            </Stack>
          </DialogContent>
          <Box sx={{ mx: 3 }}>
            <Divider />
          </Box>
          <DialogActions sx={{ p: 3 }}>
            <Button onClick={handleClose}>Cancel</Button>
          </DialogActions>
        </Dialog>
      )}
    </>
  )
}
