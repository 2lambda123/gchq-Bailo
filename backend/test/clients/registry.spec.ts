import { describe, expect, test, vi } from 'vitest'

import { listImageTags, listModelRepos } from '../../src/clients/registry.js'

const fetchMock = vi.hoisted(() => ({
  default: vi.fn(() => ({ ok: true, text: vi.fn(), json: vi.fn() })),
}))
vi.mock('node-fetch', async () => fetchMock)

describe('clients > registry', () => {
  test('listModelRepos > only returns model repos', async () => {
    const modelId = 'modelId'
    fetchMock.default.mockReturnValueOnce({
      ok: true,
      text: vi.fn(),
      json: vi.fn(() => ({ repositories: [`${modelId}/repo`, 'wrong/repo'] })),
    })
    const response = await listModelRepos('token', modelId)

    expect(fetchMock.default).toBeCalled()
    expect(fetchMock.default.mock.calls).toMatchSnapshot()
    expect(response).toStrictEqual([`${modelId}/repo`])
  })

  test('listModelRepos > cannot reach registry', async () => {
    fetchMock.default.mockRejectedValueOnce('Error')
    const response = listModelRepos('token', 'modelId')

    expect(response).rejects.toThrowError('Unable to communicate with the registry.')
  })

  test('listModelRepos > non-200 response', async () => {
    fetchMock.default.mockReturnValueOnce({
      ok: false,
      text: vi.fn(() => 'oh no something went wrong...'),
      json: vi.fn(),
    })
    const response = listModelRepos('token', 'modelId')

    expect(response).rejects.toThrowError('Non-200 response returned by the registry.')
  })

  test('listModelRepos > malformed response', async () => {
    fetchMock.default.mockReturnValueOnce({
      ok: true,
      text: vi.fn(),
      json: vi.fn(() => 'wrong'),
    })
    const response = listModelRepos('token', 'modelId')

    expect(response).rejects.toThrowError('Unrecognised response body when listing model repositories.')
  })

  test('listModelRepos > missing repositories in response', async () => {
    fetchMock.default.mockReturnValueOnce({
      ok: true,
      text: vi.fn(),
      json: vi.fn(() => ({ fake: 'info' })),
    })
    const response = listModelRepos('token', 'modelId')

    expect(response).rejects.toThrowError('Unrecognised response body when listing model repositories.')
  })

  test('listImageTags > success', async () => {
    const tags = ['tag1', 'tag2']
    fetchMock.default.mockReturnValueOnce({
      ok: true,
      text: vi.fn(),
      json: vi.fn(() => ({ tags })),
    })

    const response = await listImageTags('token', { namespace: 'modelId', image: 'image' })

    expect(fetchMock.default).toBeCalled()
    expect(fetchMock.default.mock.calls).toMatchSnapshot()
    expect(response).toStrictEqual(tags)
  })

  test('listImageTags > malformed response', async () => {
    fetchMock.default.mockReturnValueOnce({
      ok: true,
      text: vi.fn(),
      json: vi.fn(() => 'wrong'),
    })

    const response = listImageTags('token', { namespace: 'modelId', image: 'image' })

    expect(response).rejects.toThrowError('Unrecognised response body when listing image tags.')
  })

  test('listImageTags > missing repositories in response', async () => {
    fetchMock.default.mockReturnValueOnce({
      ok: true,
      text: vi.fn(),
      json: vi.fn(() => ({ fake: 'info' })),
    })

    const response = listImageTags('token', { namespace: 'modelId', image: 'image' })

    expect(response).rejects.toThrowError('Unrecognised response body when listing image tags.')
  })
})
