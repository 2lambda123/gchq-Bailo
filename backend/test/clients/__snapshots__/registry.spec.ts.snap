// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`clients > registry > listImageTags > success 1`] = `
[
  [
    "https://localhost:5000/v2/modelId/image/tags/list",
    {
      "agent": Agent {
        "_events": {
          "free": [Function],
          "newListener": [Function],
        },
        "_eventsCount": 2,
        "_maxListeners": undefined,
        "_sessionCache": {
          "list": [],
          "map": {},
        },
        "defaultPort": 443,
        "freeSockets": {},
        "keepAlive": false,
        "keepAliveMsecs": 1000,
        "maxCachedSessions": 100,
        "maxFreeSockets": 256,
        "maxSockets": Infinity,
        "maxTotalSockets": Infinity,
        "options": {
          "path": null,
          "rejectUnauthorized": false,
        },
        "protocol": "https:",
        "requests": {},
        "scheduling": "lifo",
        "sockets": {},
        "totalSocketCount": 0,
        Symbol(kCapture): false,
      },
      "headers": {
        "Authorization": "Bearer token",
      },
    },
  ],
]
`;

exports[`clients > registry > listModelRepos > only returns model repos 1`] = `
[
  [
    "https://localhost:5000/v2/_catalog?n=100&last=modelId",
    {
      "agent": Agent {
        "_events": {
          "free": [Function],
          "newListener": [Function],
        },
        "_eventsCount": 2,
        "_maxListeners": undefined,
        "_sessionCache": {
          "list": [],
          "map": {},
        },
        "defaultPort": 443,
        "freeSockets": {},
        "keepAlive": false,
        "keepAliveMsecs": 1000,
        "maxCachedSessions": 100,
        "maxFreeSockets": 256,
        "maxSockets": Infinity,
        "maxTotalSockets": Infinity,
        "options": {
          "path": null,
          "rejectUnauthorized": false,
        },
        "protocol": "https:",
        "requests": {},
        "scheduling": "lifo",
        "sockets": {},
        "totalSocketCount": 0,
        Symbol(kCapture): false,
      },
      "headers": {
        "Authorization": "Bearer token",
      },
    },
  ],
]
`;
